File: android/domain/src/main/kotlin/com/cacompte/domain/rules/GameRules.kt
package com.cacompte.domain.rules

import com.cacompte.domain.model.MatchState
import com.cacompte.domain.model.ModifierID
import com.cacompte.domain.model.RoundDraft
import com.cacompte.domain.model.ScoreEntry
import com.cacompte.domain.model.ValidationError
import com.cacompte.domain.model.ValidationResult
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.util.UUID

/** Miroir de `EndReason.swift`. */
@Serializable
enum class EndReason {
    @SerialName("scoreThreshold")
    ScoreThreshold,

    @SerialName("roundLimit")
    RoundLimit,

    @SerialName("targetReached")
    TargetReached,

    @SerialName("allSheetsComplete")
    AllSheetsComplete,

    @SerialName("elimination")
    Elimination,

    @SerialName("manualStop")
    ManualStop,
}

/** Miroir de `EndCheck` (déclaré dans `GameRules.swift`). */
sealed interface EndCheck {
    data object Continue : EndCheck

    data class FinalRound(
        val triggeredBy: UUID,
        val reason: EndReason,
    ) : EndCheck

    data class Ended(
        val reason: EndReason,
    ) : EndCheck
}

/** Miroir de `Standing.swift`. */
data class Standing(
    val participantID: UUID,
    val rank: Int,
    val score: Int,
    val sharedWith: List<UUID> = emptyList(),
)

/**
 * Miroir de `GameRules.swift` (protocole) + son extension par défaut, réunis ici — Kotlin permet
 * des corps de méthode par défaut directement dans l'interface, contrairement à Swift qui sépare
 * protocole et extension. `GenericSumRules` (`:catalog`) hérite de tout sans rien redéfinir,
 * comme son homologue Swift.
 */
interface GameRules {
    val engineID: String

    fun validate(
        draft: RoundDraft,
        state: MatchState,
        definition: GameDefinition,
    ): ValidationResult {
        val entry = definition.scoring.entry
        val errors = mutableListOf<ValidationError>()
        val warnings = mutableListOf<String>()

        for (input in draft.inputs) {
            val min = entry.min
            val max = entry.max
            if (min != null && input.rawValue < min) {
                errors +=
                    ValidationError(
                        ValidationError.Field.ParticipantField(input.participantID),
                        "Score sous le minimum autorisé ($min).",
                    )
            }
            if (max != null && input.rawValue > max) {
                errors +=
                    ValidationError(
                        ValidationError.Field.ParticipantField(input.participantID),
                        "Score au-dessus du maximum autorisé ($max).",
                    )
            }
            val warnBelow = entry.warnBelow
            if (warnBelow != null && input.rawValue < warnBelow) {
                warnings += "Score inhabituel, à vérifier."
            }
            val warnAbove = entry.warnAbove
            if (warnAbove != null && input.rawValue > warnAbove) {
                warnings += "Score inhabituel, à vérifier."
            }
        }

        return when {
            errors.isNotEmpty() -> ValidationResult.Invalid(errors)
            warnings.isNotEmpty() -> ValidationResult.Warning(warnings)
            else -> ValidationResult.Valid
        }
    }

    fun score(
        draft: RoundDraft,
        state: MatchState,
        definition: GameDefinition,
    ): List<ScoreEntry> =
        draft.inputs.map { input ->
            ScoreEntry(
                participantID = input.participantID,
                rawValue = input.rawValue,
                computedValue = input.rawValue, // identité, aucune règle appliquée
                explanation = null,
                detail = input.detail,
                modifiers = input.modifiers,
            )
        }

    fun endCheck(
        state: MatchState,
        definition: GameDefinition,
    ): EndCheck {
        val totals = state.totals()
        for (condition in definition.end.conditions) {
            val threshold = condition.resolvedValue(state.variants)
            when (condition.type) {
                EndConditionType.ScoreThreshold -> {
                    val reached =
                        state.participants.any {
                            condition.comparison.evaluate(totals[it.id] ?: 0, threshold)
                        }
                    if (reached) return EndCheck.Ended(EndReason.ScoreThreshold)
                }

                EndConditionType.RoundLimit -> {
                    if (condition.comparison.evaluate(state.rounds.size, threshold)) {
                        return EndCheck.Ended(EndReason.RoundLimit)
                    }
                }

                // Nécessitent des données propres au jeu, pas gérées génériquement.
                EndConditionType.TargetReached,
                EndConditionType.AllSheetsComplete,
                EndConditionType.Elimination,
                EndConditionType.ManualStop,
                -> Unit
            }
        }
        return EndCheck.Continue
    }

    fun standings(
        state: MatchState,
        definition: GameDefinition,
    ): List<Standing> {
        val totals = state.totals()
        val direction = definition.scoring.direction

        val groupedByTotal = state.participants.map { it.id }.groupBy { totals[it] ?: 0 }
        val orderedTotals =
            if (direction == Direction.HighestWins) {
                groupedByTotal.keys.sortedDescending()
            } else {
                groupedByTotal.keys.sorted()
            }

        val result = mutableListOf<Standing>()
        var rank = 1
        for (total in orderedTotals) {
            val tied = groupedByTotal.getValue(total)
            val subgroups = resolveTieBreakGroups(tied, definition.tieBreak, state, direction)
            for (group in subgroups) {
                for (id in group) {
                    result +=
                        Standing(
                            participantID = id,
                            rank = rank,
                            score = total,
                            sharedWith = group.filterNot { it == id },
                        )
                }
                rank += group.size
            }
        }
        return result
    }
}

/**
 * Récursion de départage — miroir de `resolveTieBreakGroups`/`groupByKey`/
 * `bestSingleRoundValue`/`roundsClosedCount` (extension de `GameRules.swift`). **Le
 * comportement `.worstSingleRound` réutilisant `bestSingleRoundValue` (le "meilleur" round quel
 * que soit le cas déclenchant, seul le sens du tri change) est un quirk du source Swift, porté
 * tel quel — les golden files (étape C) le figent, ce n'est pas une réinterprétation à corriger
 * ici.**
 */
private fun resolveTieBreakGroups(
    ids: List<UUID>,
    rules: List<TieBreakRule>,
    state: MatchState,
    direction: Direction,
): List<List<UUID>> {
    if (ids.size <= 1 || rules.isEmpty()) return listOf(ids)
    return when (rules.first()) {
        TieBreakRule.Shared -> listOf(ids)

        TieBreakRule.BestSingleRound -> {
            val values = ids.associateWith { bestSingleRoundValue(it, state, direction) }
            groupByKey(ids, values, ascending = direction != Direction.HighestWins, rules.drop(1), state, direction)
        }

        TieBreakRule.WorstSingleRound -> {
            val values = ids.associateWith { bestSingleRoundValue(it, state, direction) }
            groupByKey(ids, values, ascending = direction == Direction.HighestWins, rules.drop(1), state, direction)
        }

        TieBreakRule.FewestRoundsClosed -> {
            val values = ids.associateWith { roundsClosedCount(it, state) }
            groupByKey(ids, values, ascending = true, rules.drop(1), state, direction)
        }

        TieBreakRule.MostRoundsWon,
        TieBreakRule.LowerSecondaryScore,
        TieBreakRule.HigherSecondaryScore,
        TieBreakRule.HeadToHead,
        -> resolveTieBreakGroups(ids, rules.drop(1), state, direction)
    }
}

private fun groupByKey(
    ids: List<UUID>,
    values: Map<UUID, Int>,
    ascending: Boolean,
    rules: List<TieBreakRule>,
    state: MatchState,
    direction: Direction,
): List<List<UUID>> {
    val grouped = ids.groupBy { values.getValue(it) }
    val orderedKeys = if (ascending) grouped.keys.sorted() else grouped.keys.sortedDescending()
    return orderedKeys.flatMap { key -> resolveTieBreakGroups(grouped.getValue(key), rules, state, direction) }
}

private fun bestSingleRoundValue(
    id: UUID,
    state: MatchState,
    direction: Direction,
): Int {
    val values =
        state.rounds.mapNotNull { round ->
            round.entries.firstOrNull { it.participantID == id }?.computedValue
        }
    if (values.isEmpty()) return 0
    return if (direction == Direction.HighestWins) values.max() else values.min()
}

private fun roundsClosedCount(
    id: UUID,
    state: MatchState,
): Int =
    state.rounds.count { round ->
        round.entries
            .firstOrNull { it.participantID == id }
            ?.modifiers
            ?.contains(ModifierID.closedRound) == true
    }

File: apple/QuiMeneKit/Sources/Domain/Rules/GameRules.swift
/// Doc 04 — quatre méthodes, aucune n'a d'effet de bord, aucune ne prend de dépendance. Les
/// valeurs par défaut de l'extension couvrent entièrement `generic.sum.v1` : un jeu simple
/// n'implémente rien du tout. Un jeu avec un calcul propre (Skyjo) ne surcharge que ce qui
/// l'exige — `endCheck`/`standings` restent déclaratifs tant que `end`/`tieBreak` du JSON
/// suffisent à les décrire.
public protocol GameRules: Sendable {
  /// Identifiant du moteur, tel qu'écrit dans le JSON : `"skyjo.v1"`.
  static var engineID: String { get }

  func validate(_ draft: RoundDraft, in state: MatchState, definition: GameDefinition)
    -> ValidationResult

  func score(_ draft: RoundDraft, in state: MatchState, definition: GameDefinition) -> [ScoreEntry]

  func endCheck(_ state: MatchState, definition: GameDefinition) -> EndCheck

  func standings(_ state: MatchState, definition: GameDefinition) -> [Standing]
}

public enum EndCheck: Sendable, Equatable {
  case `continue`
  case finalRound(triggeredBy: Participant.ID, reason: EndReason)
  case ended(reason: EndReason)
}

public enum EndReason: String, Sendable, Codable, Equatable {
  case scoreThreshold
  case roundLimit
  case targetReached
  case allSheetsComplete
  case elimination
  case manualStop
}

/// Doc 03 : classement final. `sharedWith` liste les autres participants au même rang — vide
/// si le rang n'est pas partagé.
public struct Standing: Sendable, Codable, Equatable {
  public let participantID: Participant.ID
  public let rank: Int
  public let score: Int
  public let sharedWith: [Participant.ID]

  public init(
    participantID: Participant.ID, rank: Int, score: Int, sharedWith: [Participant.ID] = []
  ) {
    self.participantID = participantID
    self.rank = rank
    self.score = score
    self.sharedWith = sharedWith
  }
}

extension GameRules {
  public func validate(_ draft: RoundDraft, in state: MatchState, definition: GameDefinition)
    -> ValidationResult
  {
    let entry = definition.scoring.entry
    var errors: [ValidationError] = []
    var warnings: [String] = []

    for input in draft.inputs {
      if let min = entry.min, input.rawValue < min {
        errors.append(
          ValidationError(
            field: .participant(input.participantID),
            message: "Score sous le minimum autorisé (\(min))."
          ))
      }
      if let max = entry.max, input.rawValue > max {
        errors.append(
          ValidationError(
            field: .participant(input.participantID),
            message: "Score au-dessus du maximum autorisé (\(max))."
          ))
      }
      if let warnBelow = entry.warnBelow, input.rawValue < warnBelow {
        warnings.append("Score inhabituel, à vérifier.")
      }
      if let warnAbove = entry.warnAbove, input.rawValue > warnAbove {
        warnings.append("Score inhabituel, à vérifier.")
      }
    }

    if !errors.isEmpty { return .invalid(errors) }
    if !warnings.isEmpty { return .warning(warnings) }
    return .valid
  }

  public func score(_ draft: RoundDraft, in state: MatchState, definition: GameDefinition)
    -> [ScoreEntry]
  {
    draft.inputs.map { input in
      ScoreEntry(
        participantID: input.participantID,
        rawValue: input.rawValue,
        computedValue: input.rawValue,
        detail: input.detail,
        modifiers: input.modifiers
      )
    }
  }

  public func endCheck(_ state: MatchState, definition: GameDefinition) -> EndCheck {
    let totals = state.totals()
    for condition in definition.end.conditions {
      let threshold = condition.resolvedValue(variants: state.variants)
      switch condition.type {
      case .scoreThreshold:
        let reached = state.participants.contains {
          condition.comparison.evaluate(totals[$0.id] ?? 0, threshold)
        }
        if reached { return .ended(reason: .scoreThreshold) }
      case .roundLimit:
        if condition.comparison.evaluate(state.rounds.count, threshold) {
          return .ended(reason: .roundLimit)
        }
      case .targetReached, .allSheetsComplete, .elimination, .manualStop:
        continue  // nécessitent des données spécifiques au jeu, pas génériques
      }
    }
    return .continue
  }

  public func standings(_ state: MatchState, definition: GameDefinition) -> [Standing] {
    let totals = state.totals()
    let direction = definition.scoring.direction

    let groupedByTotal = Dictionary(grouping: state.participants.map(\.id)) { totals[$0] ?? 0 }
    let orderedTotals = groupedByTotal.keys.sorted { direction == .highestWins ? $0 > $1 : $0 < $1 }

    var result: [Standing] = []
    var rank = 1
    for total in orderedTotals {
      let tied = groupedByTotal[total] ?? []
      let subgroups = resolveTieBreakGroups(
        tied, rules: definition.tieBreak, state: state, direction: direction)
      for group in subgroups {
        for id in group {
          result.append(
            Standing(
              participantID: id,
              rank: rank,
              score: total,
              sharedWith: group.filter { $0 != id }
            ))
        }
        rank += group.count
      }
    }
    return result
  }
}

// MARK: - Départage générique (§ charte tieBreak, doc 04)

private func resolveTieBreakGroups(
  _ ids: [Participant.ID],
  rules: [TieBreakRule],
  state: MatchState,
  direction: Direction
) -> [[Participant.ID]] {
  guard ids.count > 1, let rule = rules.first else { return [ids] }

  switch rule {
  case .shared:
    return [ids]
  case .bestSingleRound:
    let values = Dictionary(
      uniqueKeysWithValues: ids.map {
        ($0, bestSingleRoundValue($0, state: state, direction: direction))
      })
    return groupByKey(
      ids, values: values, ascending: direction != .highestWins, rules: Array(rules.dropFirst()),
      state: state, direction: direction)
  case .worstSingleRound:
    let values = Dictionary(
      uniqueKeysWithValues: ids.map {
        ($0, bestSingleRoundValue($0, state: state, direction: direction))
      })
    return groupByKey(
      ids, values: values, ascending: direction == .highestWins, rules: Array(rules.dropFirst()),
      state: state, direction: direction)
  case .fewestRoundsClosed:
    let values = Dictionary(
      uniqueKeysWithValues: ids.map { ($0, roundsClosedCount($0, state: state)) })
    return groupByKey(
      ids, values: values, ascending: true, rules: Array(rules.dropFirst()), state: state,
      direction: direction)
  case .mostRoundsWon, .lowerSecondaryScore, .higherSecondaryScore, .headToHead:
    // Nécessitent des données non modélisées génériquement (classement par manche, score
    // secondaire propre au jeu) — un moteur spécifique les surchargerait s'il en avait besoin.
    return resolveTieBreakGroups(
      ids, rules: Array(rules.dropFirst()), state: state, direction: direction)
  }
}

private func groupByKey(
  _ ids: [Participant.ID],
  values: [Participant.ID: Int],
  ascending: Bool,
  rules: [TieBreakRule],
  state: MatchState,
  direction: Direction
) -> [[Participant.ID]] {
  let grouped = Dictionary(grouping: ids) { values[$0] ?? 0 }
  let orderedKeys = grouped.keys.sorted { ascending ? $0 < $1 : $0 > $1 }
  return orderedKeys.flatMap { key in
    resolveTieBreakGroups(grouped[key] ?? [], rules: rules, state: state, direction: direction)
  }
}

private func bestSingleRoundValue(_ id: Participant.ID, state: MatchState, direction: Direction)
  -> Int
{
  let values = state.rounds.compactMap { round in
    round.entries.first { $0.participantID == id }?.computedValue
  }
  guard !values.isEmpty else { return 0 }
  return direction == .highestWins ? (values.max() ?? 0) : (values.min() ?? 0)
}

private func roundsClosedCount(_ id: Participant.ID, state: MatchState) -> Int {
  state.rounds.reduce(0) { count, round in
    let closed =
      round.entries.first { $0.participantID == id }?.modifiers.contains(.closedRound) ?? false
    return count + (closed ? 1 : 0)
  }
}
